python/hail/expr/aggregators/aggregators.py

```diff
@@ -43,60 +41,75 @@ def _to_agg(x):
-                    list(builtins.map(get_ir, constructor_args)),
-                    None if init_op_args is None else list(builtins.map(get_ir, init_op_args)),
-                    signature)
-    indices, _ = unify_all(*args)
-    return construct_expr(ir, ret_type, indices, aggregations.push(Aggregation(aggregable, *args)))
+class AggFunc(object):
+    def __init__(self):
+        self._as_scan = False
+
+    @typecheck_method(name=str,
+                      aggregable=Aggregable,
+                      ret_type=hail_type,
+                      constructor_args=sequenceof(expr_any),
+                      init_op_args=nullable(sequenceof(expr_any)),
+                      f=nullable(func_spec(1, expr_any)))
+    def __call__(self, name, aggregable, ret_type, constructor_args=(), init_op_args=None, f=None):
+        args = constructor_args if init_op_args is None else constructor_args + init_op_args
+        indices, aggregations = unify_all(aggregable, *args)
+        if aggregations:
+            raise ExpressionException('Cannot aggregate an already-aggregated expression')
+        for a in args:
+            _check_agg_bindings(a)
+        _check_agg_bindings(aggregable)
+
+        def get_type(expr):
+            return expr.dtype
+
+        def get_ir(expr):
+            return expr._ir
+
+        def agg_sig(*seq_op_args):
+            return AggSignature(name,
+                                list(map(get_type, constructor_args)),
+                                None if init_op_args is None else list(map(get_type, init_op_args)),
+                                list(map(get_type, seq_op_args)))
+
+        if name == "Count":
+            def make_seq_op(_):
+                return construct_expr(SeqOp(I32(0), [], agg_sig()), None)
+            seq_op = aggregable._transformations(aggregable, make_seq_op)
+            signature = agg_sig()
+        elif f is None:
+            def make_seq_op(agg):
+                return construct_expr(SeqOp(I32(0), [get_ir(agg)], agg_sig(aggregable)), None)
+            seq_op = aggregable._transformations(aggregable, make_seq_op)
+            signature = agg_sig(aggregable)
+        else:
+            def make_seq_op(agg):
+                uid = Env.get_uid()
+                ref = construct_variable(uid, get_type(agg), agg._indices)
+                result = f(ref)
+                body = Let(uid, get_ir(agg), get_ir(result))
+                return construct_expr(SeqOp(I32(0), [get_ir(agg), body], agg_sig(agg, result)), None)
+            seq_op = aggregable._transformations(aggregable, make_seq_op)
+            signature = agg_sig(aggregable, f(aggregable))
+
+        if self._as_scan:
+            ir = ApplyScanOp(seq_op._ir,
+                             list(map(get_ir, constructor_args)),
+                             None if init_op_args is None else list(map(get_ir, init_op_args)),
+                             signature)
+            indices, _ = unify_all(aggregable, *args)
```

Comment:
scans aren't aggregations, so they don't add aggregations to the expression. The axes of the scanned expression also behave differently because if we e.g. scan along a row-indexed expression on a Table, that needs to make it into some sort of `select_rows` or `annotate_rows`, whereas an aggregation along a row-indexed expression would result in a local value.